src/test/java/org/springframework/data/redis/cache/DefaultRedisCacheWriterUnitTests.java

```diff
@@ -86,14 +121,236 @@ void getWithNullTtl() {
+		verify(this.mockCacheStatisticsCollector, times(1)).incHits(eq("TestCacheName"));
+		verifyNoMoreInteractions(this.mockConnectionFactory, this.mockConnection, this.mockCacheStatisticsCollector);
+		verifyNoMoreInteractions(mockStringCommands);
+		verifyNoInteractions(this.mockReactiveConnectionFactory, this.mockReactiveConnection);
+	}
+
+	@Test // GH-2650
+	void retrieveWithExpirationReturnsAsynchronousMonoWithValue() {
+
+		byte[] key = "TestKey".getBytes();
+
+		Duration thirtySeconds = Duration.ofSeconds(30L);
+
+		RedisStringCommands mockStringCommands = mock(RedisStringCommands.class);
+
+		doReturn(mockStringCommands).when(this.mockConnection).stringCommands();
+		doReturn("test".getBytes()).when(mockStringCommands).getEx(any(byte[].class), any(Expiration.class));
+
+		RedisCacheWriter cacheWriter = newRedisCacheWriter();
+
+		Mono<byte[]> result = cacheWriter.retrieve("TestCacheName", key, thirtySeconds);
+
+		assertThat(result).isNotNull();
+		verifyNoInteractions(this.mockCacheStatisticsCollector);
+
+		byte[] value = result.block();
+
+		assertThat(value).isNotNull();
+		assertThat(new String(value)).isEqualTo("test");
+
+		verify(this.mockConnectionFactory, times(1)).getConnection();
+		verify(this.mockConnection, times(1)).stringCommands();
+		verify(this.mockConnection, times(1)).close();
+		verify(mockStringCommands, times(1)).getEx(eq(key), eq(Expiration.from(thirtySeconds)));
+		verify(this.mockCacheStatisticsCollector, times(1)).incGets(eq("TestCacheName"));
+		verify(this.mockCacheStatisticsCollector, times(1)).incHits(eq("TestCacheName"));
+		verifyNoMoreInteractions(this.mockConnectionFactory, this.mockConnection, this.mockCacheStatisticsCollector);
+		verifyNoMoreInteractions(mockStringCommands);
+		verifyNoInteractions(this.mockReactiveConnectionFactory, this.mockReactiveConnection);
+	}
+
+	@Test // GH-2650
+	void retrieveReturnsReactiveMonoWithValue() {
+
+		byte[] key = "TestKey".getBytes();
+
+		Duration sixtySeconds = Duration.ofSeconds(60L);
+
+		RedisKeyCommands mockKeyCommands = mock(RedisKeyCommands.class);
+		ReactiveStringCommands mockStringCommands = mock(ReactiveStringCommands.class);
+
+		doReturn(mockKeyCommands).when(this.mockConnection).keyCommands();
+		doReturn(false).when(mockKeyCommands).exists(any(byte[].class));
+		doReturn(mockStringCommands).when(this.mockReactiveConnection).stringCommands();
+		doReturn(Mono.just(ByteBuffer.wrap("test".getBytes()))).when(mockStringCommands).get(any(ByteBuffer.class));
+
+		RedisCacheWriter cacheWriter = newReactiveRedisCacheWriter(sixtySeconds);
+
+		Mono<byte[]> result = cacheWriter.retrieve("TestCacheName", key);
+
+		assertThat(result).isNotNull();
+		verifyNoInteractions(this.mockCacheStatisticsCollector);
+
+		byte[] value = result.block();
+
+		assertThat(value).isNotNull();
+		assertThat(new String(value)).isEqualTo("test");
+
+		verify(this.mockReactiveConnectionFactory, times(1)).getReactiveConnection();
```

Comment:
I wonder how maintainable this will remain when we have almost 15 verifications and what net value of the test remains. 

Generally looking at the new tests, there's 90% infrastructure and only 10% of actual test code left.